src/Compiler/Microsoft.NET.Sdk.Razor.SourceGenerators/RazorSourceGenerator.cs

```diff
@@ -98,30 +99,35 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
 
             var generatedDeclarationSyntaxTrees = generatedDeclarationCode
                 .Combine(parseOptions)
-                .Select(static (pair, _) =>
+                .Select(static (pair, ct) =>
                 {
                     var (generatedDeclarationCode, parseOptions) = pair;
-                    return CSharpSyntaxTree.ParseText(generatedDeclarationCode, (CSharpParseOptions)parseOptions);
+                    return CSharpSyntaxTree.ParseText(generatedDeclarationCode, (CSharpParseOptions)parseOptions, cancellationToken: ct);
+                });
+
+            var declCompilation = generatedDeclarationSyntaxTrees
+                .Collect()
+                .Combine(compilation)
+                .Select((pair, _) =>
+                {
+                    return pair.Right.AddSyntaxTrees(pair.Left);
                 });
 
-            var tagHelpersFromCompilation = compilation
-                .Combine(generatedDeclarationSyntaxTrees.Collect())
+            var tagHelpersFromCompilation = declCompilation
                 .Combine(razorSourceGeneratorOptions)
                 .Select(static (pair, _) =>
                 {
                     RazorSourceGeneratorEventSource.Log.DiscoverTagHelpersFromCompilationStart();
 
-                    var ((compilation, generatedDeclarationSyntaxTrees), razorSourceGeneratorOptions) = pair;
+                    var (compilation, razorSourceGeneratorOptions) = pair;
 
                     var tagHelperFeature = new StaticCompilationTagHelperFeature();
                     var discoveryProjectEngine = GetDiscoveryProjectEngine(compilation.References.ToImmutableArray(), tagHelperFeature);
```

Comment:
What does this do? The variable `discoveryProjectEngine` is not used anywhere. Consider removing the variable and adding a comment why the call is needed. Same below in `tagHelpersFromReferences`.

Reply:
Yeah, this has sort of gotten weird through various round of refactoring. I've done some more refactoring to make it way more sane and easier to understand. Good call.